src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -155,149 +155,126 @@ def _is_bound_op(self, expr, bound_operator, index):
-                # fully set up.
-                return False
-            shape = symbol.datatype.shape
-            sym_maths = SymbolicMaths.get()
-            if (sym_maths.equal(shape[index].lower, shape[index].upper) and
-                    sym_maths.equal(shape[index].lower, array_dimension)):
+        access_shape = self.indices[index]
+
+        # Is this array access in the form of {UL}BOUND(array, index)?
+        if isinstance(access_shape, Range):
+            if bound_type == "upper":
+                operator = BinaryOperation.Operator.UBOUND
+                access_bound = access_shape.stop
+            else:
+                operator = BinaryOperation.Operator.LBOUND
+                access_bound = access_shape.start
+            if self._is_bound_op(access_bound, operator, index):
                 return True
+
+        # Try to compare the upper/lower bound of the array access
+        # with the upper/lower bound of the array declaration.
+
+        # Finding the array declaration is only supported for an
+        # ArrayReference at the moment.
+        # Import here to avoid circular dependence.
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.nodes.array_reference import ArrayReference
+        if not isinstance(self, ArrayReference):
+            return False
+        # pylint: enable=import-outside-toplevel
+
+        symbol = self.symbol
+        if not isinstance(symbol, DataSymbol):
+            # There is no type information for this symbol
+            # (probably because it originates from a wildcard import).
             return False
 
-        upper = array_dimension.stop
+        datatype = symbol.datatype
+        if not isinstance(datatype, ArrayType):
+            # The declaration datatype could be of UnknownFortranType
+            # if the symbol is of e.g. character type.
+            return False
+
+        # The bound of the declaration is available.
+
+        if isinstance(datatype.shape[index], ArrayType.Extent):
+            # The size is unspecified at compile-time (but is
+            # available at run-time e.g. when the size is allocated by
+            # an allocate statement.
+            return False
 
-        # Is this in the form of UBOUND(array, index)?
-        if self._is_bound_op(upper, BinaryOperation.Operator.UBOUND, index):
+        # The size of the bound is available.
+        if bound_type == "upper":
+            declaration_bound = datatype.shape[index].upper
+        else:
+            declaration_bound = datatype.shape[index].lower
+
+        # Do the bounds match?
+        if isinstance(access_shape, Range):
+            if bound_type == "upper":
+                access_bound = access_shape.stop
+            else:
+                access_bound = access_shape.start
+        else:
+            access_bound = access_shape
+        sym_maths = SymbolicMaths.get()
+        if sym_maths.equal(declaration_bound, access_bound):
```

Comment:
At this point you can just do:

    return sym_maths.equal(declaration_bound, access_bound)

I think?